tests/fmt_spec.lua

```diff
@@ -0,0 +1,46 @@
+local replace_placeholders = require('config.plugins.luasnip.utils').replace_placeholders
+
+local check = function(msg, fmt, args, expected)
+    it(msg, function()
+        local result = replace_placeholders(fmt, args)
+        assert.are.same(expected, table.concat(result))
+    end)
+end
+
+describe('replace_placeholders', function()
+    check('expands with no numbers',
+        'a{}b{}c{}d', {4, 5, 6},
+        'a4b5c6d')
+
+    check('expands with explicit numbers',
+        'a{2}b{1}c{3}d', {4, 5, 6},
+        'a5b4c6d')
+
+    check('expands with mixed numbering',
+        'a{}b{3}c{}d', {1, 2, 3, 4},
+        'a1b3c4d')
+
+    check('expands named placeholders',
+        'a{A}b{B}c{C}d', {A = 1, B = 2, C = 3},
+        'a1b2c3d')
+
+    check('expands all mixed',
+        'a {A} b {} c {3} d {} e {B} f {A} g', {1, 2, 3, 4, A = 10, B = 20},
+        'a 10 b 1 c 3 d 4 e 20 f 10 g')
```

Comment:
What happens / should happen for `'a {A} b {} c', {1, 2, B = 20, A = 10, C = 30}` ?

Reply:
This gives `a 10 b 1 c`, which is also what would happen in Python (`'a {A} b {} c'.format(1, 2, B = 20, A = 10, C = 30)` gives `'a 10 b 1 c'`). 

In current implementation this is because the numbered (`{2}`) and auto-numbered (`{}`) fields use an index that starts from 1 and is incremented on each auto-numbered placeholder or set to new value on each numbered one. Named placeholders ignore this index. What [fmtlib](https://fmt.dev/latest/syntax.html) describes and what you see in Python, is that you cannot mix auto-numbered and numbered fields, so the current implementation might be confusing on this part, e.g. in Python both `'{1} {}'` and `'{} {1}'` are errors, while this implementation treats them as `'{1} {2}'` and `'{1} {1}'` respectively.

Reply:
Makes sense to me, maybe add another example that explicitly shows this behaviour? Especially as it differs from pythons' `fmt`

Reply:
I added an example and a test for that.